### PR TITLE
plan: draw relation bewteen operators with indent

### LINKED ISSUE #6893: refine the result format of explain
The current result of `explain` is kind of confusion and hard to find the relationship between operators:
```sql
TiDB(localhost:4000) > desc select count(*) from t1 left outer join t2 on t1.a=t2.a and t1.b>10 and t2.b>20 group by t1.a;
+-----------------+-----------------+-------------------------------+------+--------------------------------------------------------------------------------------------------------+----------+
| id              | parents         | children                      | task | operator info                                                                                          | count    |
+-----------------+-----------------+-------------------------------+------+--------------------------------------------------------------------------------------------------------+----------+
| TableScan_11    |                 |                               | cop  | table:t1, range:[-inf,+inf], keep order:false                                                          | 10000.00 |
| TableReader_12  | HashLeftJoin_10 |                               | root | data:TableScan_11                                                                                      | 10000.00 |
| TableScan_13    | Selection_14    |                               | cop  | table:t2, range:[-inf,+inf], keep order:false                                                          | 10000.00 |
| Selection_14    |                 | TableScan_13                  | cop  | gt(test.t2.b, 20)                                                                                      | 3333.33  |
| TableReader_15  | HashLeftJoin_10 |                               | root | data:Selection_14                                                                                      | 3333.33  |
| HashLeftJoin_10 | HashAgg_9       | TableReader_12,TableReader_15 | root | left outer join, inner:TableReader_15, equal:[eq(test.t1.a, test.t2.a)], left cond:[gt(test.t1.b, 10)] | 10000.00 |
| HashAgg_9       |                 | HashLeftJoin_10               | root | group by:test.t1.a, funcs:count(1)                                                                     | 8000.00  |
+-----------------+-----------------+-------------------------------+------+--------------------------------------------------------------------------------------------------------+----------+
7 rows in set (0.00 sec)
```

Actually we can draw the relationship between operators by indent:
```sql
TiDB(localhost:4000) > desc select count(*) from t1 left outer join t2 on t1.a=t2.a and t1.b>10 and t2.b>20 group by t1.a;
+----------------------+------+--------------------------------------------------------------------------------------------------------+----------+
| id                   | task | operator info                                                                                          | count    |
+----------------------+------+--------------------------------------------------------------------------------------------------------+----------+
| HashAgg_9            | root | group by:test.t1.a, funcs:count(1)                                                                     | 8000.00  |
|   HashLeftJoin_10    | root | left outer join, inner:TableReader_15, equal:[eq(test.t1.a, test.t2.a)], left cond:[gt(test.t1.b, 10)] | 10000.00 |
|     TableReader_12   | root | data:TableScan_11                                                                                      | 10000.00 |
|       TableScan_11   | cop  | table:t1, range:[-inf,+inf], keep order:false                                                          | 10000.00 |
|     TableReader_15   | root | data:Selection_14                                                                                      | 3333.33  |
|       Selection_14   | cop  | gt(test.t2.b, 20)                                                                                      | 3333.33  |
|         TableScan_13 | cop  | table:t2, range:[-inf,+inf], keep order:false                                                          | 10000.00 |
+----------------------+------+--------------------------------------------------------------------------------------------------------+----------+
7 rows in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by zz-jason**
## What have you changed? (mandatory)
Fix #6893

The current result of `explain` is kind of confusion and hard to find the relationship between operators:
```sql
TiDB(localhost:4000) > desc select count(*) from t1 left outer join t2 on t1.a=t2.a and t1.b>10 and t2.b>20 group by t1.a;
+-----------------+-----------------+-------------------------------+------+--------------------------------------------------------------------------------------------------------+----------+
| id              | parents         | children                      | task | operator info                                                                                          | count    |
+-----------------+-----------------+-------------------------------+------+--------------------------------------------------------------------------------------------------------+----------+
| TableScan_11    |                 |                               | cop  | table:t1, range:[-inf,+inf], keep order:false                                                          | 10000.00 |
| TableReader_12  | HashLeftJoin_10 |                               | root | data:TableScan_11                                                                                      | 10000.00 |
| TableScan_13    | Selection_14    |                               | cop  | table:t2, range:[-inf,+inf], keep order:false                                                          | 10000.00 |
| Selection_14    |                 | TableScan_13                  | cop  | gt(test.t2.b, 20)                                                                                      | 3333.33  |
| TableReader_15  | HashLeftJoin_10 |                               | root | data:Selection_14                                                                                      | 3333.33  |
| HashLeftJoin_10 | HashAgg_9       | TableReader_12,TableReader_15 | root | left outer join, inner:TableReader_15, equal:[eq(test.t1.a, test.t2.a)], left cond:[gt(test.t1.b, 10)] | 10000.00 |
| HashAgg_9       |                 | HashLeftJoin_10               | root | group by:test.t1.a, funcs:count(1)                                                                     | 8000.00  |
+-----------------+-----------------+-------------------------------+------+--------------------------------------------------------------------------------------------------------+----------+
7 rows in set (0.00 sec)
```

Actually we can draw the relationship between operators by indent, after this PR:
```sql
TiDB(localhost:4000) > desc select count(*) from t1 left outer join t2 on t1.a=t2.a and t1.b>10 and t2.b>20 group by t1.a;
+--------------------------+------+-----------------------------------------------------------------------------------------------+----------+
| id                       | task | operator info                                                                                 | count    |
+--------------------------+------+-----------------------------------------------------------------------------------------------+----------+
| StreamAgg_12             | root | group by:test.t1.a, funcs:count(1)                                                            | 8000.00  |
| └─MergeJoin_29           | root | left outer join, left key:test.t1.a, right key:test.t2.a, left cond:[gt(cast(test.t1.b), 10)] | 10000.00 |
|   ├─IndexLookUp_18       | root | index:IndexScan_16, table:TableScan_17                                                        | 10000.00 |
|   │ ├─IndexScan_16       | cop  | table:t1, index:a, range:[<nil>,+inf], keep order:true                                        | 10000.00 |
|   │ └─TableScan_17       | cop  | table:t1, keep order:false                                                                    | 10000.00 |
|   └─Selection_19         | root | gt(cast(test.t2.b), 20)                                                                       | 8000.00  |
|     └─IndexLookUp_23     | root | index:IndexScan_21, table:TableScan_22                                                        | 10000.00 |
|       ├─IndexScan_21     | cop  | table:t2, index:a, range:[<nil>,+inf], keep order:true                                        | 10000.00 |
|       └─TableScan_22     | cop  | table:t2, keep order:false                                                                    | 10000.00 |
+--------------------------+------+-----------------------------------------------------------------------------------------------+----------+
9 rows in set (0.00 sec)
```

## What are the type of the changes (mandatory)?

- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested (mandatory)?

- unit test
- explain test

## Does this PR affect documentation (docs/docs-cn) update? (optional)

Yes

### CODE REVIEW

**Comment by ngaut** [comment]
Cool, well done.

File: plan/common_plans.go [hunk 1]
```diff
@@ -408,50 +407,44 @@ type Explain struct {
...
+func (e *Explain) prepareRootTaskInfo(p PhysicalPlan, indent string) {
+	e.prepareExplainInfo4DAGTask(p, "root", indent)
 	e.explainedPlans[p.ID()] = true
+
 	for _, child := range p.Children() {
 		if e.explainedPlans[child.ID()] {
 			continue
 		}
-		e.prepareRootTaskInfo(child.(PhysicalPlan), p.ExplainID())
+		e.prepareRootTaskInfo(child.(PhysicalPlan), indent+"  ")
```
Comment (ngaut) [inline plan/common_plans.go]: Could you extract a function like prettyIdentifier to replace the manual concat string?

File: plan/explain.go [hunk 2]
```diff
@@ -149,11 +149,13 @@ func (p *basePhysicalAgg) ExplainInfo() string {
 		fmt.Fprintf(buffer, "group by:%s, ",
 			expression.ExplainExpressionList(p.GroupByItems))
 	}
-	buffer.WriteString("funcs:")
-	for i, agg := range p.AggFuncs {
-		buffer.WriteString(aggregation.ExplainAggFunc(agg))
-		if i+1 < len(p.AggFuncs) {
-			buffer.WriteString(", ")
+	if len(p.AggFuncs) > 0 {
+		buffer.WriteString("funcs:")
```
Comment (shenli) [inline plan/explain.go]: Will there be an agg-operator without AggFuncs?

File: plan/common_plans.go (hunk 1 above)
Comment (ngaut) [reply plan/common_plans.go]: Could you use "-" instead of blank.

File: plan/explain.go (hunk 2 above)
Comment (winoros) [reply plan/explain.go]: yes it may be.

File: plan/common_plans.go (hunk 1 above)
Comment (winoros) [inline plan/common_plans.go]: Why not just `indent+="  "` out of the for loop.

Comment (winoros) [reply plan/common_plans.go]: Why not use a constant to replace the `" "`. So it will make it easier to do some change like replace the `"  "` to `"--"`

**Comment by zz-jason** [comment]
@winoros @ngaut PTAL this explain format:
```sql
+------------------------+------+-----------------------------------------------------------------------------------------------+----------+
| id                     | task | operator info                                                                                 | count    |
+------------------------+------+-----------------------------------------------------------------------------------------------+----------+
| StreamAgg_12           | root | group by:test.t1.a, funcs:count(1)                                                            | 8000.00  |
| └-MergeJoin_29         | root | left outer join, left key:test.t1.a, right key:test.t2.a, left cond:[gt(cast(test.t1.b), 10)] | 10000.00 |
|   |-IndexLookUp_18     | root | index:IndexScan_16, table:TableScan_17                                                        | 10000.00 |
|   | |-IndexScan_16     | cop  | table:t1, index:a, range:[<nil>,+inf], keep order:true                                        | 10000.00 |
|   | └-TableScan_17     | cop  | table:t1, keep order:false                                                                    | 10000.00 |
|   └-Selection_19       | root | gt(cast(test.t2.b), 20)                                                                       | 8000.00  |
|     └-IndexLookUp_23   | root | index:IndexScan_21, table:TableScan_22                                                        | 10000.00 |
|       |-IndexScan_21   | cop  | table:t2, index:a, range:[<nil>,+inf], keep order:true                                        | 10000.00 |
|       └-TableScan_22   | cop  | table:t2, keep order:false                                                                    | 10000.00 |
+------------------------+------+-----------------------------------------------------------------------------------------------+----------+
9 rows in set (0.01 sec)
```

**Comment by zz-jason** [comment]
I'll fix ci and test once we have decided the right explain format.

**Comment by shenli** [comment]
The format looks good to me.

File: plan/explain.go (hunk 2 above)
Comment (shenli) [reply plan/explain.go]: Any example?

**Comment by ngaut** [comment]
LGTM

File: plan/common_plans.go [hunk 3]
```diff
@@ -407,51 +406,97 @@ type Explain struct {
 	explainedPlans map[int]bool
 }
 
-// prepareExplainInfo4DAGTask generates the following information for every plan:
-// ["id", "parents", "task", "operator info"].
-func (e *Explain) prepareExplainInfo4DAGTask(p PhysicalPlan, taskType string, parentID string) {
-	childrenIDs := make([]string, 0, len(p.Children()))
-	for _, ch := range p.Children() {
-		childrenIDs = append(childrenIDs, ch.ExplainID())
+const (
```
Comment (winoros) [inline plan/common_plans.go]: Since this change is large enough, what about just remove these things to `explain.go`

File: plan/common_plans.go [hunk 4]
```diff
@@ -407,51 +406,97 @@ type Explain struct {
...
+	// treeBody indicates the operator tree is not finished.
+	treeBody rune = '│'
```
Comment (winoros) [inline plan/common_plans.go]: `rune` can just be omitted? `treeBody = '|'` is enough?

File: plan/common_plans.go [hunk 5]
```diff
@@ -407,51 +406,97 @@ type Explain struct {
...
+	// treeMiddleNode indicates this operator is in the middle the operator tree.
+	treeMiddleNode rune = '├'
+	// treeLastNode indicates this operator is the last child of the operator tree.
```
Comment (winoros) [inline plan/common_plans.go]: ditto. `the last child of its parent` is clearer.

File: plan/common_plans.go [hunk 6]
```diff
@@ -407,51 +406,97 @@ type Explain struct {
...
-// prepareExplainInfo4DAGTask generates the following information for every plan:
-// ["id", "parents", "task", "operator info"].
-func (e *Explain) prepareExplainInfo4DAGTask(p PhysicalPlan, taskType string, parentID string) {
-	childrenIDs := make([]string, 0, len(p.Children()))
-	for _, ch := range p.Children() {
-		childrenIDs = append(childrenIDs, ch.ExplainID())
+const (
+	// treeBody indicates the operator tree is not finished.
+	treeBody rune = '│'
+	// treeMiddleNode indicates this operator is in the middle the operator tree.
```
Comment (winoros) [inline plan/common_plans.go]: `in the middle the operator tree` is puzzled. `not the last child of its parent` is clearer.

File: plan/common_plans.go [hunk 7]
```diff
@@ -407,51 +406,97 @@ type Explain struct {
...
-		if i+1 < len(plans) {
-			parentID = plans[i+1].ExplainID()
+func (e *Explain) prettyIdentifier(id, indent string, isLastChild bool) string {
+	if len(indent) == 0 {
+		return id
+	}
+
+	// Here we attach a new node to the current operator sub-tree by changing
+	// the closest treeBody to a:
+	// 1. treeLastNode, if this node is the last node.
```
Comment (winoros) [inline plan/common_plans.go]: `last node` -> `last child` may be clearer.

File: plan/common_plans.go [hunk 8]
```diff
@@ -407,51 +406,97 @@ type Explain struct {
...
+	// need to end this sub-tree by changing the closest treeBody to a treeGap.
+	indentBytes := []rune(indent)
+	for i := len(indentBytes) - 1; i >= 0; i-- {
+		if indentBytes[i] == treeBody {
+			indentBytes[i] = treeGap
+			break
+		}
+	}
+
+	// The child belongs to a new sub-tree, which is rooted by this node.
```
Comment (winoros) [inline plan/common_plans.go]: This comment should not be here. What ever the node is the last child or not. It is always a root of its own subtree.
I think we'd better tell that `We append treeBody and treeGap to indicate that there's a subtree created`. Add add this comment at [here](https://github.com/pingcap/tidb/pull/6894/files#diff-6b4f5c526cce10da4ab025d07a2d0b32R425) too.

File: plan/common_plans.go [hunk 9]
```diff
@@ -407,51 +406,97 @@ type Explain struct {
...
 }
 
-// prepareCopTaskInfo generates explain information for cop-tasks.
-// Only PhysicalTableReader, PhysicalIndexReader and PhysicalIndexLookUpReader have cop-tasks currently.
-func (e *Explain) prepareCopTaskInfo(plans []PhysicalPlan) {
-	for i, p := range plans {
-		var parentID string
-		if i+1 < len(plans) {
-			parentID = plans[i+1].ExplainID()
+func (e *Explain) prettyIdentifier(id, indent string, isLastChild bool) string {
```
Comment (winoros) [inline plan/common_plans.go]: Add comment header for this method.

**Comment by winoros** [comment]
The dot format explain is meaningless since the new explain have the similar function with it?

**Comment by shenli** [comment]
The dot format is more pretty.

**Comment by zz-jason** [comment]
We still need the dot format, once the query relates more than 10 operators, the row format is is too complex to clearly show the relationship between operators, dot format is more clear than row format in this situation.

File: plan/common_plans.go (hunk 3 above)
Comment (zz-jason) [reply plan/common_plans.go]: It's better to be remained on this file for now. `explain.go` is for operators to print their explain info. we can do these code clean work in another pr.

**Comment by zz-jason** [comment]
@winoros @shenli PTAL

**Comment by coocood** [comment]
@zz-jason 
the operator info of `IndexLookUP` can be removed, `index:IndexScan_14, table:TableScan_15` can be inferred by the lines.

**Pull request merged**